mlflow/langchain/_langchain_autolog.py

```diff
@@ -238,40 +249,121 @@ def _update_langchain_model_config(model):
+    """
+    Inject list of callbacks into the original callbacks.
+    For RunnableConfig, the callbacks is defined as List[BaseCallbackHandler] or BaseCallbackManager
+    https://github.com/langchain-ai/langchain/blob/ed980601e1c630f996aabf85df5cb26178e53099/libs/core/langchain_core/callbacks/base.py#L636
+
+    Args:
+        original_callbacks: List[BaseCallbackHandler] or BaseCallbackManager
+        new_callbacks: List[BaseCallbackHandler]
+    """
+    if _is_base_callback_manager(original_callbacks):
+        added_callbacks = []
+        try:
+            original_callbacks, added_callbacks = _inject_callbacks_into_base_callback_manager(
+                original_callbacks, new_callbacks
+            )
+            yield original_callbacks
+        except Exception as e:
+            _logger.warning(
+                f"Autologging with tracing failed due to error: {e}",
+                exc_info=_logger.isEnabledFor(logging.DEBUG),
+            )
+        finally:
+            # avoid modifying original callbackmanager
+            for callback in added_callbacks:
+                original_callbacks.remove_handler(callback)
+
+    elif _is_list_of_base_callback_handlers(original_callbacks):
+        yield _inject_callbacks_into_list_of_base_callback_handlers(
+            original_callbacks, new_callbacks
+        )
+
+    else:
+        _logger.warning(
+            f"Unsupported callbacks type `{original_callbacks}` for autologging with tracing."
+        )
+        yield original_callbacks
+
+
+@contextlib.contextmanager
+def _inject_callbacks_to_config_list(config_list, new_callbacks):
+    """
+    Inject new_callbacks into each config in the config_list.
+    """
+    added_callbacks_list = {}
+    original_callbacks_list = {}
+    for i, config in enumerate(config_list):
+        original_callbacks = config.get("callbacks") or []
+        if _is_base_callback_manager(original_callbacks):
+            original_callbacks, added_callbacks = _inject_callbacks_into_base_callback_manager(
+                original_callbacks, new_callbacks
+            )
+            config["callbacks"] = original_callbacks
+            added_callbacks_list[i] = added_callbacks
+        elif _is_list_of_base_callback_handlers(original_callbacks):
+            config["callbacks"] = _inject_callbacks_into_list_of_base_callback_handlers(
+                original_callbacks, new_callbacks
+            )
+            original_callbacks_list[i] = original_callbacks
+    try:
+        yield config_list
+    finally:
+        for i, config in enumerate(config_list):
+            # original callback must be baseCallbackManager
+            if i in added_callbacks_list:
+                for callback in added_callbacks_list[i]:
+                    config["callbacks"].remove_handler(callback)
+            # original callback must be list of baseCallbackHandlers
+            else:
+                config["callbacks"] = original_callbacks_list[i] or None
```

Comment:
Does this line insert `callbacks` when it doesn't exist?


```python
my_config = [{}, {}]

model.invoke(..., config=my_config)

print(my_config)  # -> [{"callbacks": None}, {"callbacks": None}]
```

Reply:
Yes if it's not existing then we set it to None, I think it's fine because the default value is None.